AnnabelleAB/CS5500-assignment3
# Implement Chat Item List Ordering and Error Handling

### Git Issue
![image](https://github.com/AnnabelleAB/CS5500-assignment3/assets/49429865/14933db7-d4ba-41c5-addb-c64edff21db3)


### Description
This PR introduces a server-side mechanism to maintain an ordered list of chat items based on their timestamps. It ensures that chat items are retrieved and displayed in chronological order. Additionally, this PR includes the implementation of API endpoints on the server to receive chat items and the development of error-handling mechanisms for scenarios like failed message transmission, network issues, and server errors.



### Changes
- Modified `DocumentServer.ts` to include a sorting mechanism that orders chat items by their timestamp upon retrieval.
- Updated the POST endpoint in `DocumentServer.ts` to handle the reception of chat items with necessary fields such as `user`, `message`, and `timestamp`.
- Implemented comprehensive error handling in `DocumentServer.ts` for the POST and GET endpoints to manage various failure scenarios gracefully.
- Added tests in `DocumentServerTester.ts` using `axios` to simulate client requests and validate the functionality of chat item ordering and error handling.

### Tests
- **Chat Item Ordering**: Tests have been added to ensure that chat items are returned in the correct order by timestamp.
- **Error Handling**: Tests have been included to verify that the server responds with appropriate error messages and status codes when encountering bad requests or server-side issues.
- **Endpoint Functionality**: Additional tests check the successful creation of new chat items via the POST endpoint and the correct retrieval of existing items via the GET endpoint.

### How to Test
- Ensure the server is running locally on the specified port.
- Run `ts-node src/Server/DocumentServerTester.ts` to execute the test suite.
- Verify that all tests pass without errors.

### Additional Notes
- The server port and endpoint paths used in tests should match the actual server configuration.
- Error handling tests are designed to confirm that the server does not crash and provides meaningful error messages to the client.

### Terminal Test Log 
![image](https://github.com/AnnabelleAB/CS5500-assignment3/assets/49429865/1c7da8c5-1d41-45c6-ace8-81f14906b230)
